// @bnb-chain/greenfield-cosmos-types/greenfield/sp/authz.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DepositAuthorization = exports.protobufPackage = void 0;
/* eslint-disable */
const coin_1 = require("../../cosmos/base/v1beta1/coin");
const _m0 = __importStar(require("protobufjs/minimal"));
const helpers_1 = require("../../helpers");
exports.protobufPackage = "greenfield.sp";
function createBaseDepositAuthorization() {
    return {
        maxDeposit: undefined,
        spAddress: ""
    };
}
exports.DepositAuthorization = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.maxDeposit !== undefined) {
            coin_1.Coin.encode(message.maxDeposit, writer.uint32(10).fork()).ldelim();
        }
        if (message.spAddress !== "") {
            writer.uint32(18).string(message.spAddress);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseDepositAuthorization();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.maxDeposit = coin_1.Coin.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.spAddress = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            maxDeposit: (0, helpers_1.isSet)(object.maxDeposit) ? coin_1.Coin.fromJSON(object.maxDeposit) : undefined,
            spAddress: (0, helpers_1.isSet)(object.spAddress) ? String(object.spAddress) : ""
        };
    },
    toJSON(message) {
        const obj = {};
        message.maxDeposit !== undefined && (obj.maxDeposit = message.maxDeposit ? coin_1.Coin.toJSON(message.maxDeposit) : undefined);
        message.spAddress !== undefined && (obj.spAddress = message.spAddress);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseDepositAuthorization();
        message.maxDeposit = object.maxDeposit !== undefined && object.maxDeposit !== null ? coin_1.Coin.fromPartial(object.maxDeposit) : undefined;
        message.spAddress = object.spAddress ?? "";
        return message;
    },
    fromSDK(object) {
        return {
            maxDeposit: object.max_deposit ? coin_1.Coin.fromSDK(object.max_deposit) : undefined,
            spAddress: object?.sp_address
        };
    },
    toSDK(message) {
        const obj = {};
        message.maxDeposit !== undefined && (obj.max_deposit = message.maxDeposit ? coin_1.Coin.toSDK(message.maxDeposit) : undefined);
        obj.sp_address = message.spAddress;
        return obj;
    }
};
