// @bnb-chain/greenfield-cosmos-types/greenfield/sp/params.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Params = exports.protobufPackage = void 0;
/* eslint-disable */
const _m0 = __importStar(require("protobufjs/minimal"));
const helpers_1 = require("../../helpers");
exports.protobufPackage = "greenfield.sp";
function createBaseParams() {
    return {
        depositDenom: "",
        minDeposit: "",
        secondarySpStorePriceRatio: ""
    };
}
exports.Params = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.depositDenom !== "") {
            writer.uint32(10).string(message.depositDenom);
        }
        if (message.minDeposit !== "") {
            writer.uint32(18).string(message.minDeposit);
        }
        if (message.secondarySpStorePriceRatio !== "") {
            writer.uint32(26).string(message.secondarySpStorePriceRatio);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseParams();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.depositDenom = reader.string();
                    break;
                case 2:
                    message.minDeposit = reader.string();
                    break;
                case 3:
                    message.secondarySpStorePriceRatio = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            depositDenom: (0, helpers_1.isSet)(object.depositDenom) ? String(object.depositDenom) : "",
            minDeposit: (0, helpers_1.isSet)(object.minDeposit) ? String(object.minDeposit) : "",
            secondarySpStorePriceRatio: (0, helpers_1.isSet)(object.secondarySpStorePriceRatio) ? String(object.secondarySpStorePriceRatio) : ""
        };
    },
    toJSON(message) {
        const obj = {};
        message.depositDenom !== undefined && (obj.depositDenom = message.depositDenom);
        message.minDeposit !== undefined && (obj.minDeposit = message.minDeposit);
        message.secondarySpStorePriceRatio !== undefined && (obj.secondarySpStorePriceRatio = message.secondarySpStorePriceRatio);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseParams();
        message.depositDenom = object.depositDenom ?? "";
        message.minDeposit = object.minDeposit ?? "";
        message.secondarySpStorePriceRatio = object.secondarySpStorePriceRatio ?? "";
        return message;
    },
    fromSDK(object) {
        return {
            depositDenom: object?.deposit_denom,
            minDeposit: object?.min_deposit,
            secondarySpStorePriceRatio: object?.secondary_sp_store_price_ratio
        };
    },
    toSDK(message) {
        const obj = {};
        obj.deposit_denom = message.depositDenom;
        obj.min_deposit = message.minDeposit;
        obj.secondary_sp_store_price_ratio = message.secondarySpStorePriceRatio;
        return obj;
    }
};
