// @bnb-chain/greenfield-cosmos-types/cosmos/oracle/v1/event.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventPackageClaim = exports.protobufPackage = void 0;
/* eslint-disable */
const helpers_1 = require("../../../helpers");
const _m0 = __importStar(require("protobufjs/minimal"));
exports.protobufPackage = "cosmos.oracle.v1";
function createBaseEventPackageClaim() {
    return {
        srcChainId: 0,
        destChainId: 0,
        channelId: 0,
        packageType: 0,
        receiveSequence: helpers_1.Long.UZERO,
        sendSequence: helpers_1.Long.ZERO,
        crash: false,
        errorMsg: "",
        relayerFee: "",
        ackRelayerFee: ""
    };
}
exports.EventPackageClaim = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.srcChainId !== 0) {
            writer.uint32(8).uint32(message.srcChainId);
        }
        if (message.destChainId !== 0) {
            writer.uint32(16).uint32(message.destChainId);
        }
        if (message.channelId !== 0) {
            writer.uint32(24).uint32(message.channelId);
        }
        if (message.packageType !== 0) {
            writer.uint32(32).uint32(message.packageType);
        }
        if (!message.receiveSequence.isZero()) {
            writer.uint32(40).uint64(message.receiveSequence);
        }
        if (!message.sendSequence.isZero()) {
            writer.uint32(48).int64(message.sendSequence);
        }
        if (message.crash === true) {
            writer.uint32(56).bool(message.crash);
        }
        if (message.errorMsg !== "") {
            writer.uint32(66).string(message.errorMsg);
        }
        if (message.relayerFee !== "") {
            writer.uint32(74).string(message.relayerFee);
        }
        if (message.ackRelayerFee !== "") {
            writer.uint32(82).string(message.ackRelayerFee);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseEventPackageClaim();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.srcChainId = reader.uint32();
                    break;
                case 2:
                    message.destChainId = reader.uint32();
                    break;
                case 3:
                    message.channelId = reader.uint32();
                    break;
                case 4:
                    message.packageType = reader.uint32();
                    break;
                case 5:
                    message.receiveSequence = reader.uint64();
                    break;
                case 6:
                    message.sendSequence = reader.int64();
                    break;
                case 7:
                    message.crash = reader.bool();
                    break;
                case 8:
                    message.errorMsg = reader.string();
                    break;
                case 9:
                    message.relayerFee = reader.string();
                    break;
                case 10:
                    message.ackRelayerFee = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            srcChainId: (0, helpers_1.isSet)(object.srcChainId) ? Number(object.srcChainId) : 0,
            destChainId: (0, helpers_1.isSet)(object.destChainId) ? Number(object.destChainId) : 0,
            channelId: (0, helpers_1.isSet)(object.channelId) ? Number(object.channelId) : 0,
            packageType: (0, helpers_1.isSet)(object.packageType) ? Number(object.packageType) : 0,
            receiveSequence: (0, helpers_1.isSet)(object.receiveSequence) ? helpers_1.Long.fromValue(object.receiveSequence) : helpers_1.Long.UZERO,
            sendSequence: (0, helpers_1.isSet)(object.sendSequence) ? helpers_1.Long.fromValue(object.sendSequence) : helpers_1.Long.ZERO,
            crash: (0, helpers_1.isSet)(object.crash) ? Boolean(object.crash) : false,
            errorMsg: (0, helpers_1.isSet)(object.errorMsg) ? String(object.errorMsg) : "",
            relayerFee: (0, helpers_1.isSet)(object.relayerFee) ? String(object.relayerFee) : "",
            ackRelayerFee: (0, helpers_1.isSet)(object.ackRelayerFee) ? String(object.ackRelayerFee) : ""
        };
    },
    toJSON(message) {
        const obj = {};
        message.srcChainId !== undefined && (obj.srcChainId = Math.round(message.srcChainId));
        message.destChainId !== undefined && (obj.destChainId = Math.round(message.destChainId));
        message.channelId !== undefined && (obj.channelId = Math.round(message.channelId));
        message.packageType !== undefined && (obj.packageType = Math.round(message.packageType));
        message.receiveSequence !== undefined && (obj.receiveSequence = (message.receiveSequence || helpers_1.Long.UZERO).toString());
        message.sendSequence !== undefined && (obj.sendSequence = (message.sendSequence || helpers_1.Long.ZERO).toString());
        message.crash !== undefined && (obj.crash = message.crash);
        message.errorMsg !== undefined && (obj.errorMsg = message.errorMsg);
        message.relayerFee !== undefined && (obj.relayerFee = message.relayerFee);
        message.ackRelayerFee !== undefined && (obj.ackRelayerFee = message.ackRelayerFee);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseEventPackageClaim();
        message.srcChainId = object.srcChainId ?? 0;
        message.destChainId = object.destChainId ?? 0;
        message.channelId = object.channelId ?? 0;
        message.packageType = object.packageType ?? 0;
        message.receiveSequence = object.receiveSequence !== undefined && object.receiveSequence !== null ? helpers_1.Long.fromValue(object.receiveSequence) : helpers_1.Long.UZERO;
        message.sendSequence = object.sendSequence !== undefined && object.sendSequence !== null ? helpers_1.Long.fromValue(object.sendSequence) : helpers_1.Long.ZERO;
        message.crash = object.crash ?? false;
        message.errorMsg = object.errorMsg ?? "";
        message.relayerFee = object.relayerFee ?? "";
        message.ackRelayerFee = object.ackRelayerFee ?? "";
        return message;
    },
    fromSDK(object) {
        return {
            srcChainId: object?.src_chain_id,
            destChainId: object?.dest_chain_id,
            channelId: object?.channel_id,
            packageType: object?.package_type,
            receiveSequence: object?.receive_sequence,
            sendSequence: object?.send_sequence,
            crash: object?.crash,
            errorMsg: object?.error_msg,
            relayerFee: object?.relayer_fee,
            ackRelayerFee: object?.ack_relayer_fee
        };
    },
    toSDK(message) {
        const obj = {};
        obj.src_chain_id = message.srcChainId;
        obj.dest_chain_id = message.destChainId;
        obj.channel_id = message.channelId;
        obj.package_type = message.packageType;
        obj.receive_sequence = message.receiveSequence;
        obj.send_sequence = message.sendSequence;
        obj.crash = message.crash;
        obj.error_msg = message.errorMsg;
        obj.relayer_fee = message.relayerFee;
        obj.ack_relayer_fee = message.ackRelayerFee;
        return obj;
    }
};
