// @bnb-chain/greenfield-cosmos-types/greenfield/resource/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.resourceTypeToJSON = exports.resourceTypeFromJSON = exports.ResourceTypeSDKType = exports.ResourceType = exports.protobufPackage = void 0;
/* eslint-disable */
exports.protobufPackage = "greenfield.resource";
var ResourceType;
(function (ResourceType) {
    ResourceType[ResourceType["RESOURCE_TYPE_UNSPECIFIED"] = 0] = "RESOURCE_TYPE_UNSPECIFIED";
    ResourceType[ResourceType["RESOURCE_TYPE_BUCKET"] = 1] = "RESOURCE_TYPE_BUCKET";
    ResourceType[ResourceType["RESOURCE_TYPE_OBJECT"] = 2] = "RESOURCE_TYPE_OBJECT";
    ResourceType[ResourceType["RESOURCE_TYPE_GROUP"] = 3] = "RESOURCE_TYPE_GROUP";
    ResourceType[ResourceType["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(ResourceType = exports.ResourceType || (exports.ResourceType = {}));
exports.ResourceTypeSDKType = ResourceType;
function resourceTypeFromJSON(object) {
    switch (object) {
        case 0:
        case "RESOURCE_TYPE_UNSPECIFIED":
            return ResourceType.RESOURCE_TYPE_UNSPECIFIED;
        case 1:
        case "RESOURCE_TYPE_BUCKET":
            return ResourceType.RESOURCE_TYPE_BUCKET;
        case 2:
        case "RESOURCE_TYPE_OBJECT":
            return ResourceType.RESOURCE_TYPE_OBJECT;
        case 3:
        case "RESOURCE_TYPE_GROUP":
            return ResourceType.RESOURCE_TYPE_GROUP;
        case -1:
        case "UNRECOGNIZED":
        default:
            return ResourceType.UNRECOGNIZED;
    }
}
exports.resourceTypeFromJSON = resourceTypeFromJSON;
function resourceTypeToJSON(object) {
    switch (object) {
        case ResourceType.RESOURCE_TYPE_UNSPECIFIED:
            return "RESOURCE_TYPE_UNSPECIFIED";
        case ResourceType.RESOURCE_TYPE_BUCKET:
            return "RESOURCE_TYPE_BUCKET";
        case ResourceType.RESOURCE_TYPE_OBJECT:
            return "RESOURCE_TYPE_OBJECT";
        case ResourceType.RESOURCE_TYPE_GROUP:
            return "RESOURCE_TYPE_GROUP";
        case ResourceType.UNRECOGNIZED:
        default:
            return "UNRECOGNIZED";
    }
}
exports.resourceTypeToJSON = resourceTypeToJSON;
