// @bnb-chain/greenfield-cosmos-types/greenfield/payment/params.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Params = exports.protobufPackage = void 0;
/* eslint-disable */
const helpers_1 = require("../../helpers");
const _m0 = __importStar(require("protobufjs/minimal"));
exports.protobufPackage = "greenfield.payment";
function createBaseParams() {
    return {
        reserveTime: helpers_1.Long.UZERO,
        paymentAccountCountLimit: helpers_1.Long.UZERO,
        forcedSettleTime: helpers_1.Long.UZERO,
        maxAutoForceSettleNum: helpers_1.Long.UZERO,
        feeDenom: "",
        validatorTaxRate: ""
    };
}
exports.Params = {
    encode(message, writer = _m0.Writer.create()) {
        if (!message.reserveTime.isZero()) {
            writer.uint32(8).uint64(message.reserveTime);
        }
        if (!message.paymentAccountCountLimit.isZero()) {
            writer.uint32(16).uint64(message.paymentAccountCountLimit);
        }
        if (!message.forcedSettleTime.isZero()) {
            writer.uint32(24).uint64(message.forcedSettleTime);
        }
        if (!message.maxAutoForceSettleNum.isZero()) {
            writer.uint32(32).uint64(message.maxAutoForceSettleNum);
        }
        if (message.feeDenom !== "") {
            writer.uint32(42).string(message.feeDenom);
        }
        if (message.validatorTaxRate !== "") {
            writer.uint32(50).string(message.validatorTaxRate);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseParams();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.reserveTime = reader.uint64();
                    break;
                case 2:
                    message.paymentAccountCountLimit = reader.uint64();
                    break;
                case 3:
                    message.forcedSettleTime = reader.uint64();
                    break;
                case 4:
                    message.maxAutoForceSettleNum = reader.uint64();
                    break;
                case 5:
                    message.feeDenom = reader.string();
                    break;
                case 6:
                    message.validatorTaxRate = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            reserveTime: (0, helpers_1.isSet)(object.reserveTime) ? helpers_1.Long.fromValue(object.reserveTime) : helpers_1.Long.UZERO,
            paymentAccountCountLimit: (0, helpers_1.isSet)(object.paymentAccountCountLimit) ? helpers_1.Long.fromValue(object.paymentAccountCountLimit) : helpers_1.Long.UZERO,
            forcedSettleTime: (0, helpers_1.isSet)(object.forcedSettleTime) ? helpers_1.Long.fromValue(object.forcedSettleTime) : helpers_1.Long.UZERO,
            maxAutoForceSettleNum: (0, helpers_1.isSet)(object.maxAutoForceSettleNum) ? helpers_1.Long.fromValue(object.maxAutoForceSettleNum) : helpers_1.Long.UZERO,
            feeDenom: (0, helpers_1.isSet)(object.feeDenom) ? String(object.feeDenom) : "",
            validatorTaxRate: (0, helpers_1.isSet)(object.validatorTaxRate) ? String(object.validatorTaxRate) : ""
        };
    },
    toJSON(message) {
        const obj = {};
        message.reserveTime !== undefined && (obj.reserveTime = (message.reserveTime || helpers_1.Long.UZERO).toString());
        message.paymentAccountCountLimit !== undefined && (obj.paymentAccountCountLimit = (message.paymentAccountCountLimit || helpers_1.Long.UZERO).toString());
        message.forcedSettleTime !== undefined && (obj.forcedSettleTime = (message.forcedSettleTime || helpers_1.Long.UZERO).toString());
        message.maxAutoForceSettleNum !== undefined && (obj.maxAutoForceSettleNum = (message.maxAutoForceSettleNum || helpers_1.Long.UZERO).toString());
        message.feeDenom !== undefined && (obj.feeDenom = message.feeDenom);
        message.validatorTaxRate !== undefined && (obj.validatorTaxRate = message.validatorTaxRate);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseParams();
        message.reserveTime = object.reserveTime !== undefined && object.reserveTime !== null ? helpers_1.Long.fromValue(object.reserveTime) : helpers_1.Long.UZERO;
        message.paymentAccountCountLimit = object.paymentAccountCountLimit !== undefined && object.paymentAccountCountLimit !== null ? helpers_1.Long.fromValue(object.paymentAccountCountLimit) : helpers_1.Long.UZERO;
        message.forcedSettleTime = object.forcedSettleTime !== undefined && object.forcedSettleTime !== null ? helpers_1.Long.fromValue(object.forcedSettleTime) : helpers_1.Long.UZERO;
        message.maxAutoForceSettleNum = object.maxAutoForceSettleNum !== undefined && object.maxAutoForceSettleNum !== null ? helpers_1.Long.fromValue(object.maxAutoForceSettleNum) : helpers_1.Long.UZERO;
        message.feeDenom = object.feeDenom ?? "";
        message.validatorTaxRate = object.validatorTaxRate ?? "";
        return message;
    },
    fromSDK(object) {
        return {
            reserveTime: object?.reserve_time,
            paymentAccountCountLimit: object?.payment_account_count_limit,
            forcedSettleTime: object?.forced_settle_time,
            maxAutoForceSettleNum: object?.max_auto_force_settle_num,
            feeDenom: object?.fee_denom,
            validatorTaxRate: object?.validator_tax_rate
        };
    },
    toSDK(message) {
        const obj = {};
        obj.reserve_time = message.reserveTime;
        obj.payment_account_count_limit = message.paymentAccountCountLimit;
        obj.forced_settle_time = message.forcedSettleTime;
        obj.max_auto_force_settle_num = message.maxAutoForceSettleNum;
        obj.fee_denom = message.feeDenom;
        obj.validator_tax_rate = message.validatorTaxRate;
        return obj;
    }
};
