// @bnb-chain/greenfield-cosmos-types/greenfield/payment/payment_account.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PaymentAccount = exports.protobufPackage = void 0;
/* eslint-disable */
const _m0 = __importStar(require("protobufjs/minimal"));
const helpers_1 = require("../../helpers");
exports.protobufPackage = "greenfield.payment";
function createBasePaymentAccount() {
    return {
        addr: "",
        owner: "",
        refundable: false
    };
}
exports.PaymentAccount = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.addr !== "") {
            writer.uint32(10).string(message.addr);
        }
        if (message.owner !== "") {
            writer.uint32(18).string(message.owner);
        }
        if (message.refundable === true) {
            writer.uint32(24).bool(message.refundable);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePaymentAccount();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.addr = reader.string();
                    break;
                case 2:
                    message.owner = reader.string();
                    break;
                case 3:
                    message.refundable = reader.bool();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            addr: (0, helpers_1.isSet)(object.addr) ? String(object.addr) : "",
            owner: (0, helpers_1.isSet)(object.owner) ? String(object.owner) : "",
            refundable: (0, helpers_1.isSet)(object.refundable) ? Boolean(object.refundable) : false
        };
    },
    toJSON(message) {
        const obj = {};
        message.addr !== undefined && (obj.addr = message.addr);
        message.owner !== undefined && (obj.owner = message.owner);
        message.refundable !== undefined && (obj.refundable = message.refundable);
        return obj;
    },
    fromPartial(object) {
        const message = createBasePaymentAccount();
        message.addr = object.addr ?? "";
        message.owner = object.owner ?? "";
        message.refundable = object.refundable ?? false;
        return message;
    },
    fromSDK(object) {
        return {
            addr: object?.addr,
            owner: object?.owner,
            refundable: object?.refundable
        };
    },
    toSDK(message) {
        const obj = {};
        obj.addr = message.addr;
        obj.owner = message.owner;
        obj.refundable = message.refundable;
        return obj;
    }
};
