// @bnb-chain/greenfield-cosmos-types/greenfield/bridge/params.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Params = exports.protobufPackage = void 0;
/* eslint-disable */
const _m0 = __importStar(require("protobufjs/minimal"));
const helpers_1 = require("../../helpers");
exports.protobufPackage = "greenfield.bridge";
function createBaseParams() {
    return {
        transferOutRelayerFee: "",
        transferOutAckRelayerFee: ""
    };
}
exports.Params = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.transferOutRelayerFee !== "") {
            writer.uint32(10).string(message.transferOutRelayerFee);
        }
        if (message.transferOutAckRelayerFee !== "") {
            writer.uint32(18).string(message.transferOutAckRelayerFee);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseParams();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.transferOutRelayerFee = reader.string();
                    break;
                case 2:
                    message.transferOutAckRelayerFee = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            transferOutRelayerFee: (0, helpers_1.isSet)(object.transferOutRelayerFee) ? String(object.transferOutRelayerFee) : "",
            transferOutAckRelayerFee: (0, helpers_1.isSet)(object.transferOutAckRelayerFee) ? String(object.transferOutAckRelayerFee) : ""
        };
    },
    toJSON(message) {
        const obj = {};
        message.transferOutRelayerFee !== undefined && (obj.transferOutRelayerFee = message.transferOutRelayerFee);
        message.transferOutAckRelayerFee !== undefined && (obj.transferOutAckRelayerFee = message.transferOutAckRelayerFee);
        return obj;
    },
    fromPartial(object) {
        const message = createBaseParams();
        message.transferOutRelayerFee = object.transferOutRelayerFee ?? "";
        message.transferOutAckRelayerFee = object.transferOutAckRelayerFee ?? "";
        return message;
    },
    fromSDK(object) {
        return {
            transferOutRelayerFee: object?.transfer_out_relayer_fee,
            transferOutAckRelayerFee: object?.transfer_out_ack_relayer_fee
        };
    },
    toSDK(message) {
        const obj = {};
        obj.transfer_out_relayer_fee = message.transferOutRelayerFee;
        obj.transfer_out_ack_relayer_fee = message.transferOutAckRelayerFee;
        return obj;
    }
};
