// @bnb-chain/greenfield-cosmos-types/greenfield/permission/params.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Params = exports.protobufPackage = void 0;
/* eslint-disable */
const helpers_1 = require("../../helpers");
const _m0 = __importStar(require("protobufjs/minimal"));
exports.protobufPackage = "greenfield.permission";
function createBaseParams() {
    return {
        maximumStatementsNum: helpers_1.Long.UZERO,
        maximumGroupNum: helpers_1.Long.UZERO
    };
}
exports.Params = {
    encode(message, writer = _m0.Writer.create()) {
        if (!message.maximumStatementsNum.isZero()) {
            writer.uint32(8).uint64(message.maximumStatementsNum);
        }
        if (!message.maximumGroupNum.isZero()) {
            writer.uint32(16).uint64(message.maximumGroupNum);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseParams();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.maximumStatementsNum = reader.uint64();
                    break;
                case 2:
                    message.maximumGroupNum = reader.uint64();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            maximumStatementsNum: (0, helpers_1.isSet)(object.maximumStatementsNum) ? helpers_1.Long.fromValue(object.maximumStatementsNum) : helpers_1.Long.UZERO,
            maximumGroupNum: (0, helpers_1.isSet)(object.maximumGroupNum) ? helpers_1.Long.fromValue(object.maximumGroupNum) : helpers_1.Long.UZERO
        };
    },
    toJSON(message) {
        const obj = {};
        message.maximumStatementsNum !== undefined && (obj.maximumStatementsNum = (message.maximumStatementsNum || helpers_1.Long.UZERO).toString());
        message.maximumGroupNum !== undefined && (obj.maximumGroupNum = (message.maximumGroupNum || helpers_1.Long.UZERO).toString());
        return obj;
    },
    fromPartial(object) {
        const message = createBaseParams();
        message.maximumStatementsNum = object.maximumStatementsNum !== undefined && object.maximumStatementsNum !== null ? helpers_1.Long.fromValue(object.maximumStatementsNum) : helpers_1.Long.UZERO;
        message.maximumGroupNum = object.maximumGroupNum !== undefined && object.maximumGroupNum !== null ? helpers_1.Long.fromValue(object.maximumGroupNum) : helpers_1.Long.UZERO;
        return message;
    },
    fromSDK(object) {
        return {
            maximumStatementsNum: object?.maximum_statements_num,
            maximumGroupNum: object?.maximum_group_num
        };
    },
    toSDK(message) {
        const obj = {};
        obj.maximum_statements_num = message.maximumStatementsNum;
        obj.maximum_group_num = message.maximumGroupNum;
        return obj;
    }
};
